containers/src/Data/Set/Internal.hs

```diff
@@ -2192,6 +2208,271 @@ instance Monoid (MergeSet a) where
+--
+-- @since FIXME
+runWhenMissing :: WhenMissing f a -> a -> f Bool
+runWhenMissing = missingElem
+
+-- | A tactic for dealing with elements present in both sets in 'merge' or
+-- 'mergeA'.
+--
+-- A tactic of type @WhenMatched f a@ is an abstract representation of a
+-- function of type @a -> f Bool@.
+--
+-- @since FIXME
+newtype WhenMatched f a = WhenMatched { matchedElem :: a -> f Bool }
+
+-- | A tactic for dealing with elements present in both sets in 'merge'.
+--
+-- A tactic of type @SimpleWhenMatched a@ is an abstract representation of a
+-- function of type @a -> Bool@.
+--
+-- @since FIXME
+type SimpleWhenMatched = WhenMatched Identity
+
+-- | Along with 'filterAMatched', witnesses the isomorphism between
+-- @WhenMatched f a@ and @a -> f Bool@.
+--
+-- @since FIXME
+runWhenMatched :: WhenMatched f a -> a -> f Bool
+runWhenMatched = matchedElem
+
+-- | When an element is found in both sets, choose whether to keep the element
+-- in the merged set.
+--
+-- @since FIXME
+filterMatched :: Applicative f => (a -> Bool) -> WhenMatched f a
+filterMatched f = WhenMatched (pure . f)
+{-# INLINE filterMatched #-}
+
+-- | When an element is found in both sets, choose whether to keep the element
+-- in the merged set.
+--
+-- @since FIXME
+filterAMatched :: (a -> f Bool) -> WhenMatched f a
+filterAMatched = WhenMatched
+
+-- | Drop all the elements that are missing from the other set.
+--
+-- @
+-- dropMissing :: 'SimpleWhenMissing' a
+-- @
+--
+-- > dropMissing = filterMissing (\_ -> False)
+--
+-- but @dropMissing@ is more efficient.
+--
+-- @since FIXME
+dropMissing :: Applicative f => WhenMissing f a
+dropMissing = WhenMissing
+  { missingSubtree = \_ -> pure Tip
+  , missingElem = \_ -> pure False
+  }
+{-# INLINE dropMissing #-}
+
+-- | Preserve the elements that are missing from the other set.
+--
+-- @
+-- preserveMissing :: 'SimpleWhenMissing' a
+-- @
+--
+-- > preserveMissing = filterMissing (\_ -> True)
+--
+-- but @preserveMissing@ is more efficient.
+--
+-- @since FIXME
+preserveMissing :: Applicative f => WhenMissing f a
+preserveMissing = WhenMissing
+  { missingSubtree = pure
+  , missingElem = \_ -> pure True
+  }
+{-# INLINE preserveMissing #-}
+
+-- | Filter the elements that are missing from the other set.
+--
+-- @
+-- filterMissing :: (a -> Bool) -> 'SimpleWhenMissing' a
+-- @
+--
+-- @since FIXME
+filterMissing :: Applicative f => (a -> Bool) -> WhenMissing f a
+filterMissing f = WhenMissing
+  { missingSubtree = pure . filter f
+  , missingElem = pure . f
+  }
+{-# INLINE filterMissing #-}
+
+-- | Filter the elements that are missing from the other set using some
+-- 'Applicative' action.
+--
+-- @since FIXME
+filterAMissing :: Applicative f => (a -> f Bool) -> WhenMissing f a
+filterAMissing f = WhenMissing
+  { missingSubtree = filterA f
+  , missingElem = f
+  }
+{-# INLINE filterAMissing #-}
+
+filterA :: Applicative f => (a -> f Bool) -> Set a -> f (Set a)
+filterA f = go
+  where
```

Comment:
Do we want to start with this? I think it should help a little in general and a lot for some functors.

```haskell
go t@(Bin 1 x _ _) = finish <$> f x
  where
    finish False = Tip
    finish True = t
```

The extra sharing here makes me think we probably want to share any unchanged subtrees. We don't even need to use pointer equality for that, though doing so might turn out to be fastest.

Reply:
Thanks, updated to use `ptrEq` as we do in `filter`. I think it makes sense to keep them consistent.

Reply:
Agreed. I still think you should consider special-casing a size of 1. We inherently know its subtrees don't change, plus we avoid the `liftA3`.